34.md

```diff
@@ -0,0 +1,84 @@
+NIP-34
+======
+
+`git` stuff
+-----------
+
+`draft` `optional`
+
+This NIP defines all the ways code collaboration using and adjacent to [`git`](https://git-scm.com/) can be done using Nostr.
+
+## Repository announcements
+
+Git repositories are hosted in Git-enabled servers, but their existence can be announced using Nostr events, as well as their willingness to receive patches, bug reports and comments in general.
+
+```jsonc
+{
+  "kind": 30617,
+  "content": "",
+  "tags": [
+    ["d", "<repo-id>"],
+    ["name", "<human-readable project name>"],
+    ["description", "brief human-readable project description>"],
+    ["head", "<sha1-hex>"], // the current repository commit head, as hex
```

Comment:
what's this for? doesn't git handle this?

Reply:
The idea was that people could optionally provide this to prevent untrusted git servers from serving bad code under their name.

Reply:
the git url is authoritative though, someone could spoof the url, the hash would be different, and users would be none the wiser. maybe I'm still misunderstanding.

Reply:
I had this idea too. This obviously relies on a new 30617 event being issued every time a maintainer pushed to the git server.

Each maintainer would need to use a nostr git-remote-helper that proxies interaction with the git server and issues the event automatically. 

My implementation was a little more involved and allowed tracking additional branches. It also listed the tip's branch ancestors so that clients could tell how far behind they are without talking to the git server.
```
[
 ["main", "<commid_id>", "<parent commit_id shorthand>", "<grandparent commit_id shorthand>" ....]
]
```
I ultimately concluded that most projects that care about this attack vector are signing their commits which, assuming the signatures are being validated, addresses the concerns without having to introduce a new tool.

Reply:
Makes sense. I've removed it.